platform/mbed_wait_api_rtos.cpp

```diff
@@ -23,26 +23,45 @@
 #include "rtos/rtos.h"
 #include "platform/mbed_critical.h"
 #include "platform/mbed_power_mgmt.h"
+#include "platform/mbed_error.h"
 
 void wait(float s)
 {
-    wait_us(s * 1000000.0f);
+    if (s >= 0.01f) {
+        wait_ms(s * 1000.0f);
+    } else {
+        wait_us(s * 1000000.0f);
+    }
 }
 
+/*  The millisec value specifies the number of timer ticks and is therefore an upper bound.
+ *  The exact time delay depends on the actual time elapsed since the last timer tick.
+ *  The actual time delay may be up to one timer tick less.
+ */
 void wait_ms(int ms)
 {
-    wait_us(ms * 1000);
+    if (core_util_are_interrupts_enabled()) {
+        ThisThread::sleep_for(ms);
+    } else {
+        MBED_WARNING(MBED_MAKE_ERROR(MBED_MODULE_PLATFORM, MBED_ERROR_INVALID_OPERATION),
+                    "Deprecated behavior: milli-sec delay should not be used in interrupt.\n");
+        wait_us(ms * 1000);
+    }
 }
 
 void wait_us(int us)
 {
+    if (us > 10000) {
+        MBED_WARNING(MBED_MAKE_ERROR(MBED_MODULE_PLATFORM, MBED_ERROR_UNKNOWN),
+                     "wait_us blocks deep sleep, wait_ms recommended for long delays\n");
+    }
```

Comment:
I was suggesting dropping the sleep functionality from this entirely - restoring it to be the same as the non-RTOS version, hence the message above. What's your thought on that?

Reply:
Yes can do that 👍

Reply:
Well, that's kind of my design point. I also prefer yielding to other threads, but I want to define whether a call yields or not - and we would already have the yielding call, which is `wait_ms`.

This is a call which already spins for up to a millisecond to do its business, so is already inherently thread-unfriendly - I want people to be using `wait_ms` to avoid any spin.

Uses of this should be rare. (Can we do a quick audit - how many actual direct wait_us users are there in tree? Most examples I can think of are GPIO "reset" pulses on init.)

And there are some occasions where it is important not to yield, so I'd like to have an API explicitly defined as that, not dependent on parameter.

Reply:
Lets get more opinions over here
@TacoGrandeTX @c1728p9  @geky @bulislaw @stevew817 @NeilMacMullen - Please share your thoughts for `wait_us`

Reply:
Oh, and this behaviour was my original reason to have my value warning above - we're saying "you do realise you're spinning for 10ms, right?". Not just the deep sleep issue. 

I also think you've got a better chance of getting the time you were hoping each time with the spin than if you yield, where you might suddenly be hit with a full 10ms timeslice from another running thread.

Basically anyone wanting sub-millisecond resolution is having special requirements that can't really be catered for if you start dragging the RTOS into it. My feeling is that the current `wait_us` behaviour is purely to handle the `wait_ms` and `wait` built on top of it, rather than something I'd want an explicit `wait_us` to do.

Reply:
I understand the intent here to have `wait_us` as spinning only, but to be true I am not sure of all use cases it is used for currently, hence will like to know from others as well. 

Also agree that current `wait_us` behavior is to handle `wait` built on top of it.

Reply:
FWIW, just searched through the in-tree source, and I can see around a 100 calls to `wait_us`, but I can't immediately identify any calls that might ask for >=1000us, except for `wait_ms` and `wait` themselves.

Actually, no, I see a couple in one of the "onboard_modem_api.c" files, in "wiggle the reset line" code. They're a full second, so would hit the error/warning.

Reply:
Its `wait_ms` - Should not have issue, unless called from interrupt
https://github.com/ARMmbed/mbed-os/blob/master/targets/TARGET_NXP/TARGET_LPC176X/TARGET_UBLOX_C027/onboard_modem_api.c#L32

Reply:
One more thought, we can change `wait_us` to be always spinning, and update `wait` to current behavior of `wait_us`